lib/EarlyFunctionAnalysis/DetectABI.cpp

```diff
@@ -500,6 +528,117 @@ DetectABI::tryGetRegisterState(model::Register::Values RegisterValue,
   return std::nullopt;
 }
 
+using RegistersSet = std::set<model::Register::Values>;
+using RegistersLattice = SetUnionLattice<RegistersSet>;
+
+struct LatticeElement {
+  RegistersLattice Arguments;
+  RegistersLattice Returns;
+};
+
+struct Node {
+  enum class AnalysisType {
+    UAOF,
+    RAOFC,
+    URVOF,
+    URVOFC,
+  };
+
+  enum class ResultType {
+    Arguments,
+    Returns
+  };
+
+  MetaAddress Entry;
+  std::variant<AnalysisType, ResultType> Type;
+
+  explicit Node(MetaAddress Entry, AnalysisType AT) : Entry{Entry}, Type{AT} {}
+  explicit Node(MetaAddress Entry, ResultType RT) : Entry{Entry}, Type{RT} {}
+};
+
+using NodeType = ForwardNode<Node>;
+
+struct FunctionNodeSet {
+  using Ptr = std::unique_ptr<NodeType>;
+  Ptr UAOF;
+  Ptr RAOFC;
+  Ptr URVOF;
+  Ptr URVOFC;
+
+  Ptr Args;
+  Ptr Rets;
+};
+
+void DetectABI::constructInterproceduralGraph() {
+  using InterproceduralNode = ForwardNode<Node>;
+  using GraphType = GenericGraph<InterproceduralNode>;
+
+  std::map<MetaAddress, FunctionNodeSet> M;
+
+  auto GetNode = [&M](MetaAddress Entry) -> FunctionNodeSet & {
+    auto It = M.find(Entry);
+    if (It != M.end()) {
+      return It->second;
+    }
+
+    FunctionNodeSet &Nodes = M[Entry];
+
+    Nodes
+      .Args = std::make_unique<ForwardNode<Node>>(Entry,
+                                                  Node::ResultType::Arguments);
+    Nodes.Rets = std::make_unique<ForwardNode<Node>>(Entry,
+        Node::ResultType::Returns);
+
+    Nodes.UAOF = std::make_unique<ForwardNode<Node>>(Entry,
+        Node::AnalysisType::UAOF);
+    Nodes.RAOFC = std::make_unique<ForwardNode<Node>>(Entry,
+        Node::AnalysisType::RAOFC);
+    Nodes.URVOF = std::make_unique<ForwardNode<Node>>(Entry,
+        Node::AnalysisType::URVOF);
+    Nodes.URVOFC = std::make_unique<ForwardNode<Node>>(Entry,
+        Node::AnalysisType::URVOFC);
+
+    Nodes.UAOF->addSuccessor(Nodes.Args.get());
```

Comment:
1. The API of `GenericGraph` was not really designed to add edges on nodes that are not attached to the graph, it might work but it's not the intended usage.
2. I don't understand why we're adding edges between UAOF and Args? What about all the other edges?